### PR TITLE
*: fix TidbMonitor for DmCluster with TLS enabled; add some e2e test cases

### LINKED ISSUE #3987: TidbMonitor can't monitor DmCluster with TLS enabled
## Bug Report

**What version of Kubernetes are you using?**

**What version of TiDB Operator are you using?**

1.2.0-beta.2

**What storage classes exist in the Kubernetes cluster and what are used for PD/TiKV pods?**

**What's the status of the TiDB cluster pods?**

**What did you do?**

use TidbMonitor to monitor a DmCluster with TLS enabled

**What did you expect to see?**

works normally

**What did you see instead?**

no targets show in Prometheus and got some logs for Prometheus as bellow

```
level=info ts=2021-05-21T06:32:06.647Z caller=main.go:646 msg="Server is ready to receive web requests."
level=error ts=2021-05-21T06:32:11.648Z caller=manager.go:188 component="scrape manager" msg="error creating new scrape pool" err="error creating HTTP client: unable to load specified CA cert /var/lib/dm-cluster-client-tls/secret_dm-cluster-9981_tls-dm-dm-client-secret_ca.crt: open /var/lib/dm-cluster-client-tls/secret_dm-cluster-9981_tls-dm-dm-client-secret_ca.crt: no such file or directory" scrape_pool=dm-cluster-9981-tls-dm-dm-master
level=error ts=2021-05-21T06:32:11.648Z caller=manager.go:188 component="scrape manager" msg="error creating new scrape pool" err="error creating HTTP client: unable to load specified CA cert /var/lib/dm-cluster-client-tls/secret_dm-cluster-9981_tls-dm-dm-client-secret_ca.crt: open /var/lib/dm-cluster-client-tls/secret_dm-cluster-9981_tls-dm-dm-client-secret_ca.crt: no such file or directory" scrape_pool=dm-cluster-9981-tls-dm-dm-worker
```

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve?

- fix #3987
- add some e2e test cases for DM part of #2868
  - non-root run
  - monitor
  - random kill/HA

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Fix the issue that TidbMonitor may not work correctly for DmCluster with TLS enabled
```

### CODE REVIEW

**Comment by mikechengwei** [comment]
I think we only need one asset secret instead of multiple secrets. There will be other tls requirements later

**Comment by csuzhangxc** [comment]
> I think we only need one asset secret instead of multiple secrets. There will be other tls requirements later

@DanielZhangQD WDYT

**Comment by mikechengwei** [comment]
ns + secret name can be a unique path id.

**Comment by mikechengwei** [comment]
You need to understand the design idea of ​​an asset secret manage all secret. I can’t see any bugs in this PR.
About dm tls path configuration ,refer to https://github.com/pingcap/tidb-operator/blob/c5c4d59aad5d424db8a33ab175651864a1d3efd6/pkg/monitor/monitor/template.go#L425.

**Comment by csuzhangxc** [comment]
> You need to understand the design idea of ​​an asset secret manage all secret. I can’t see any bugs in this PR.
> About dm tls path configuration ,refer to
> 
> https://github.com/pingcap/tidb-operator/blob/c5c4d59aad5d424db8a33ab175651864a1d3efd6/pkg/monitor/monitor/template.go#L425
> 
> .

previously, controller-manager added all TLS certs (both TiDB and DM) into one secret and mount it under `util.ClusterAssetsTLSPath`), but TidbMonitor tries to read TLS certs for DM from `util.DMClusterClientTLSPath` (which is a different path from `util.ClusterAssetsTLSPath`).

so, to fix this problem, two methods exists:
- one as you mentioned above (`I think we only need one asset secret instead of multiple secrets. There will be other tls requirements later`)
- the other is to use different secrets for different resources, as this PR did

**Comment by mikechengwei** [comment]
> > You need to understand the design idea of ​​an asset secret manage all secret. I can’t see any bugs in this PR.
> > About dm tls path configuration ,refer to
> > https://github.com/pingcap/tidb-operator/blob/c5c4d59aad5d424db8a33ab175651864a1d3efd6/pkg/monitor/monitor/template.go#L425
> > 
> > .
> 
> previously, controller-manager added all TLS certs (both TiDB and DM) into one secret and mount it under `util.ClusterAssetsTLSPath`), but TidbMonitor tries to read TLS certs for DM from `util.DMClusterClientTLSPath` (which is a different path from `util.ClusterAssetsTLSPath`).
> 
> so, to fix this problem, two methods exist:
> 
> * one as you mentioned above (`I think we only need one asset secret instead of multiple secrets. There will be other tls requirements later`)
> * another is to use different secrets for different resources, as this PR did

We can just   change util.DMClusterClientTLSPath to  util.ClusterAssetsTLSPath.

**Comment by mikechengwei** [comment]
image
@cs

**Comment by csuzhangxc** [comment]
> image
> @csuzhangxc

I know.  that's what I've commented above "one as you mentioned above".

so, if you prefer to use "one asset secret instead of multiple secrets", we can just "change util.DMClusterClientTLSPath to util.ClusterAssetsTLSPath". 

@DanielZhangQD what do you prefer?

**Comment by DanielZhangQD** [comment]
@csuzhangxc We can just use one asset store and change the cert path to the correct one.

**Comment by csuzhangxc** [comment]
@DanielZhangQD @mikechengwei changed to use only one assert store in https://github.com/pingcap/tidb-operator/pull/3991/commits/6d5edf9aa046faab445e07d834357a0d6f625321

File: pkg/monitor/monitor/monitor_manager.go [hunk 1]
```diff
@@ -685,6 +685,7 @@ func (m *MonitorManager) syncAssetSecret(monitor *v1alpha1.TidbMonitor, store *S
 		klog.Errorf("Fail to sync tm[%s/%s]'s secret assets, err: %v", ns, name, err)
 		return err
 	}
+
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/monitor/monitor/monitor_manager.go]: reverted

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Comment by mikechengwei** [comment]
LGTM

**Review by july2993 — APPROVED** [review]

**Pull request merged**